agent-control/src/packages/extract.rs

```diff
@@ -0,0 +1,198 @@
+use flate2::read::GzDecoder;
+use std::fs::File;
+use std::path::Path;
+use tar::Archive;
+use thiserror::Error;
+use tracing::{debug, instrument};
+use zip::ZipArchive;
+
+#[derive(Debug, Error)]
+#[error("extract error: {0}")]
+pub struct ExtractError(String);
+
+pub enum SupportedArchive {
+    TarGz,
+    Zip,
+}
+
+#[instrument(skip_all, fields(archive_path = %archive_path.to_string_lossy()),name = "extracting_archive")]
+pub fn extract(
+    archive_path: &Path,
+    destination_path: &Path,
+    archive_type: SupportedArchive,
+) -> Result<(), ExtractError> {
+    match archive_type {
+        SupportedArchive::TarGz => extract_tar_gz(archive_path, destination_path),
+        SupportedArchive::Zip => extract_zip(archive_path, destination_path),
+    }
+}
+
+/// Extracts a tar.gz archive located at `archive_path` into the directory at `destination_path`.
+/// This operation is relatively sensitive in that it will not write files outside of the path specified by dst.
+/// Files in the archive which have a '..' in their path are skipped during the unpacking process.
+fn extract_tar_gz(archive_path: &Path, destination_path: &Path) -> Result<(), ExtractError> {
+    debug!("Extracting tar.gz archive to '{:?}'", destination_path);
+
+    let tar_gz = File::open(archive_path)
+        .map_err(|e| ExtractError(format!("opening tar.gz file: {:?}", e.to_string())))?;
+    let tar = GzDecoder::new(tar_gz);
+    Archive::new(tar)
+        .unpack(destination_path)
+        .map_err(|e| ExtractError(format!("extracting tar.gz file: {:?}", e.to_string())))
+}
+
+/// Extracts a zip archive located at `zip_path` into the directory at `destination`.
+/// Extraction is not atomic. If an error is encountered, some of the files may be left on disk.
+/// Extract a Zip archive into a directory, overwriting files if they already exist. Paths are sanitized with ZipFile::enclosed_name.
+fn extract_zip(zip_path: &Path, destination: &Path) -> Result<(), ExtractError> {
+    debug!("Extracting zip archive to '{:?}'", destination);
+
+    let file = File::open(zip_path)
+        .map_err(|e| ExtractError(format!("opening zip file: {:?}", e.to_string())))?;
```

Comment:
```suggestion
        .map_err(|e| ExtractError(format!("opening zip file: {e})))?;
```

Why using the debug?